src/GafferUI/StandardNodeGadget.cpp

```diff
@@ -1158,10 +1192,20 @@ bool StandardNodeGadget::dragEnter( GadgetPtr gadget, const DragDropEvent &event
 	{
 		// Display the labels for all the compatible nodules so the
 		// user can see their options.
+		IECore::ConstBoolDataPtr inputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( node(), g_inputNoduleLabelsVisibleKey );
+		IECore::ConstBoolDataPtr outputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( node(), g_outputNoduleLabelsVisibleKey );
+
 		for( StandardNodule::RecursiveIterator it( this ); !it.done(); ++it )
 		{
-			(*it)->setLabelVisible( canConnect( event, it->get() ) );
+			if(
+				( (*it)->plug()->direction() == Plug::Direction::In && !inputVisibleData ) ||
+				( (*it)->plug()->direction() == Plug::Direction::Out && !outputVisibleData )
+			)
+			{
+				(*it)->setLabelVisible( canConnect( event, it->get() ) );
+			}
 		}
+
```

Comment:
As for `enter()`, I think we'd be better off using the original code here. Currently we get no feedback at all when the metadata is specified (with either true or false). When the labels aren't visible by default, we need to see them now to have a chance of making a sensible connection. And when they are visible by default, it's useful to  hide the non-compatible ones so we get a clearer picture of the relevant options.

Reply:
Reverted and squashed into https://github.com/GafferHQ/gaffer/pull/6135/commits/7ec405e6f2e7789e74760eaf2695c9d70f2b811a